fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -1175,6 +1185,51 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
             !processIntent(intent)
     }
 
+    private fun showFullscreenMessageSurveyIfNeeded(context: Context) {
+        val messagingStorage = context.components.analytics.messagingStorage
+        val messages = runBlockingIncrement { messagingStorage.getMessages() }
+        val nextMessage =
+            messagingStorage.getNextMessage(FenixMessageSurfaceId.SURVEY, messages)
+                ?: return
+
+        val fenixNimbusMessagingController = FenixNimbusMessagingController(messagingStorage)
+        val researchSurfaceDialogFragment = ResearchSurfaceDialogFragment.newInstance(
+            keyMessageText = nextMessage.data.text,
+            keyAcceptButtonText = nextMessage.data.buttonLabel,
+            keyDismissButtonText = null,
+        )
+
+        researchSurfaceDialogFragment.onAccept = {
+            val actionUri = fenixNimbusMessagingController.processMessageActionToUri(nextMessage)
+            processIntent(Intent(Intent.ACTION_VIEW, actionUri))
+
+            components.appStore.dispatch(AppAction.MessagingAction.MessageClicked(nextMessage))
+        }
+
+        researchSurfaceDialogFragment.onDismiss = {
+            components.appStore.dispatch(AppAction.MessagingAction.MessageDismissed(nextMessage))
+        }
+
+        lifecycleScope.launch(Main) {
+            researchSurfaceDialogFragment.showNow(
+                supportFragmentManager,
+                ResearchSurfaceDialogFragment.FRAGMENT_TAG,
+            )
+        }
+
+        // Update message as displayed.
+        val currentBootUniqueIdentifier = BootUtils.getBootIdentifier(context)
+        val updatedMessage =
+            fenixNimbusMessagingController.updateMessageAsDisplayed(
+                nextMessage,
+                currentBootUniqueIdentifier,
+            )
+
+        runBlockingIncrement { fenixNimbusMessagingController.onMessageDisplayed(updatedMessage) }
+
```

Comment:
I think that defining `showFullscreenMessageSurveyIfNeeded` as `suspend` would avoid the need of using `runBlockingIncrement` here, but maybe I'm missing something.
Does it make sense to you?

Reply:
Good catch.  Thanks I'll update.